src/naemon/utils.c

```diff
@@ -535,14 +545,62 @@ int daemon_init(void)
 			return (ERROR);
 		}
 	}
-	if ((pid = (int)fork()) < 0)
+
+	/*
+	 * set up a pipe used for sending a message from the child to the parent
+	 * when initialization is complete.
+	 */
+	if (pipe(upipe_fd) < 0) {
+		nm_log(NSLOG_RUNTIME_ERROR, "Failed to set up unnamned pipe: %s",
+			strerror(errno));
 		return (ERROR);
+	}
 
-	/* parent process goes away.. */
-	else if (pid != 0)
-		exit(OK);
+	/*
+	 * as a parent and if the child dies during initialization, we need to exit
+	 * so we don't get stuck in an endless wait for input.
+	 */
+	if (signal(SIGCHLD, sighandler) == SIG_ERR) {
+		nm_log(NSLOG_RUNTIME_ERROR, "Failed to set up daemonization signal: %s",
+			strerror(errno));
+		return (ERROR);
+	}
 
-	/* child continues... */
+	if ((pid = (int)fork()) < 0) {
+		nm_log(NSLOG_RUNTIME_ERROR, "Unable to fork out the daemon process: %s",
+			strerror(errno));
+		return (ERROR);
+	} else if (pid != 0) {
+		/* parent stops - when child is done, we exit here */
+		int returnCode = OK;
+		if (close(upipe_fd[PIPE_WRITE]) < 0) {
+			nm_log(NSLOG_RUNTIME_ERROR, "Unable to close parent write end: %s",
+				strerror(errno));
+			returnCode = EXIT_FAILURE;
+		}
+		if (read(upipe_fd[PIPE_READ], &returnCode, sizeof(int)) < 0) {
+			nm_log(NSLOG_RUNTIME_ERROR, "Unable to read from pipe: %s",
+				strerror(errno));
+			returnCode = EXIT_FAILURE;
+		}
+		if (close(upipe_fd[PIPE_READ]) < 0) {
+			nm_log(NSLOG_RUNTIME_ERROR, "Unable to close parent read end: %s",
+				strerror(errno));
+			returnCode = EXIT_FAILURE;
+		}
+		if (returnCode != 0) {
```

Comment:
I think that we should compare with either `OK` or `EXIT_FAILURE` as those are the values that are assigned to `returnCode`